### PR TITLE
Implemented a different speed normalization technique for diagonals

### PR DESCRIPTION
**Pull request by MatteoPiovanelli-Laser**
Per discussion on discord on September 27, I'm finally able to PR the code that uses the intersection between an ellipse and a straight line to normalize move speed.

image
from the image:
`a` is `speed_x`
`b` is `speed_y`
The red segment is the amount of movement that will be computed for the stage. `q` is used in the drawing instead of the greek letter theta for the angle.
The equation of an ellipse is `(x/a)^2 + (y/b)^2 = 1`. The equation of the straight line from which we drew the red segment is `y = mx`. The slope `m` is found as the ratio `ydist / xdist`.
The proposed method computes the length of the red segment, as the intersection of those two equations.

<details>
<summary>Some computed results follow, to compare the current method from the ags4 branch with the proposed one.</summary>
`speed_x`, `speed_y`, `xdist`, `ydist` are the inputs to the method.
`useMoveSpeed`, `newxmove`, `newymove` are the values compute using the existing method (on ags4 branch).
`ellipse_move_speed`, `ellipse_x`, `ellispe_y` are the values computed by the proposed method.

|speed_x|speed_y|xdist|ydist|useMoveSpeed    |newxmove        |newymove         |ellipse_move_speed|ellipse_x        |ellipse_y       |
|-------|-------|-----|-----|----------------|----------------|-----------------|------------------|-----------------|----------------|
|1      |1      |1    |1    |1               |0.707106781186548|0.707106781186548|1                 |0.707106781186548|0.707106781186548|
|2      |1      |1    |1    |1.5             |1.06066017177982|1.06066017177982 |1.26491106406735  |0.894427190999916|0.894427190999916|
|3      |1      |1    |1    |2               |1.4142135623731 |1.4142135623731  |1.34164078649987  |0.948683298050514|0.948683298050514|
|4      |1      |1    |1    |2.5             |1.76776695296637|1.76776695296637 |1.37198868114007  |0.970142500145332|0.970142500145332|
|5      |1      |1    |1    |3               |2.12132034355964|2.12132034355964 |1.38675049056307  |0.98058067569092 |0.98058067569092|
|6      |1      |1    |1    |3.5             |2.47487373415292|2.47487373415292 |1.39497166492583  |0.986393923832144|0.986393923832144|
|7      |1      |1    |1    |4               |2.82842712474619|2.82842712474619 |1.4               |0.989949493661167|0.989949493661167|
|1      |1      |1    |2    |1               |0.447213595499958|0.894427190999916|1                 |0.447213595499958|0.894427190999916|
|2      |1      |1    |2    |1.33333333333333|0.596284793999944|1.19256958799989 |1.08465228909328  |0.485071250072666|0.970142500145332|
|3      |1      |1    |2    |1.66666666666667|0.74535599249993|1.49071198499986 |1.10282193314071  |0.493196961916072|0.986393923832144|
|4      |1      |1    |2    |2               |0.894427190999916|1.78885438199983 |1.10940039245046  |0.496138938356834|0.992277876713668|
|5      |1      |1    |2    |2.33333333333333|1.0434983894999 |2.0869967789998  |1.11248539872496  |0.497518595104995|0.995037190209989|
|6      |1      |1    |2    |2.66666666666667|1.19256958799989|2.38513917599978 |1.11417202906231  |0.49827287912244 |0.99654575824488|
|7      |1      |1    |2    |3               |1.34164078649987|2.68328157299975 |1.11519272878504  |0.498729349915368|0.997458699830735|
|1      |1      |1    |3    |1               |0.316227766016838|0.948683298050514|1                 |0.316227766016838|0.948683298050514|
|2      |1      |1    |3    |1.25            |0.395284707521047|1.18585412256314 |1.03975048982007  |0.328797974610715|0.986393923832144|
|3      |1      |1    |3    |1.5             |0.474341649025257|1.42302494707577 |1.04764544365437  |0.33129457822454 |0.993883734673619|
|4      |1      |1    |3    |1.75            |0.553398590529466|1.6601957715884  |1.05045146287778  |0.33218191941496 |0.99654575824488|
|5      |1      |1    |3    |2               |0.632455532033676|1.89736659610103 |1.05175790477918  |0.33259505261887 |0.997785157856609|
|6      |1      |1    |3    |2.25            |0.711512473537885|2.13453742061366 |1.05246962316844  |0.332820117735138|0.998460353205412|
|7      |1      |1    |3    |2.5             |0.790569415042095|2.37170824512628 |1.05289946579333  |0.332956045908146|0.998868137724438|
|1      |1      |2    |1    |1               |0.894427190999916|0.447213595499958|1                 |0.894427190999916|0.447213595499958|
|2      |1      |2    |1    |1.66666666666667|1.49071198499986|0.74535599249993 |1.58113883008419  |1.4142135623731  |0.707106781186548|
|3      |1      |2    |1    |2.33333333333333|2.0869967789998 |1.0434983894999  |1.86052101883813  |1.66410058867569 |0.832050294337844|
|4      |1      |2    |1    |3               |2.68328157299975|1.34164078649987 |2                 |1.78885438199983 |0.894427190999916|
|5      |1      |2    |1    |3.66666666666667|3.27956636699969|1.63978318349985 |2.0761369963435   |1.85695338177052 |0.928476690885259|
|6      |1      |2    |1    |4.33333333333333|3.87585116099964|1.93792558049982 |2.12132034355964  |1.89736659610103 |0.948683298050514|
|7      |1      |2    |1    |5               |4.47213595499958|2.23606797749979 |2.15003290891883  |1.92304789528165 |0.961523947640823|
|1      |1      |3    |1    |1               |0.948683298050514|0.316227766016838|1                 |0.948683298050514|0.316227766016838|
|2      |1      |3    |1    |1.75            |1.6601957715884 |0.553398590529466|1.75411603861406  |1.66410058867569 |0.554700196225229|
|3      |1      |3    |1    |2.5             |2.37170824512628|0.790569415042095|2.23606797749979  |2.12132034355964 |0.707106781186548|
|4      |1      |3    |1    |3.25            |3.08322071866417|1.02774023955472 |2.5298221281347   |2.4              |0.8             |
|5      |1      |3    |1    |4               |3.79473319220206|1.26491106406735 |2.7116307227332   |2.57247877713763 |0.857492925712544|
|6      |1      |3    |1    |4.75            |4.50624566573994|1.50208188857998 |2.82842712474619  |2.68328157299975 |0.894427190999916|
|7      |1      |3    |1    |5.5             |5.21775813927783|1.73925271309261 |2.9065917948809   |2.75743509005417 |0.919145030018058|
|1      |1      |1    |1    |1               |0.707106781186548|0.707106781186548|1                 |0.707106781186548|0.707106781186548|
|1      |2      |1    |1    |1.5             |1.06066017177982|1.06066017177982 |1.26491106406735  |0.894427190999916|0.894427190999916|
|1      |3      |1    |1    |2               |1.4142135623731 |1.4142135623731  |1.34164078649987  |0.948683298050514|0.948683298050514|
|1      |4      |1    |1    |2.5             |1.76776695296637|1.76776695296637 |1.37198868114007  |0.970142500145332|0.970142500145332|
|1      |5      |1    |1    |3               |2.12132034355964|2.12132034355964 |1.38675049056307  |0.98058067569092 |0.98058067569092|
|1      |6      |1    |1    |3.5             |2.47487373415292|2.47487373415292 |1.39497166492583  |0.986393923832144|0.986393923832144|
|1      |7      |1    |1    |4               |2.82842712474619|2.82842712474619 |1.4               |0.989949493661167|0.989949493661167|
|1      |1      |1    |2    |1               |0.447213595499958|0.894427190999916|1                 |0.447213595499958|0.894427190999916|
|1      |2      |1    |2    |1.66666666666667|0.74535599249993|1.49071198499986 |1.58113883008419  |0.707106781186548|1.4142135623731 |
|1      |3      |1    |2    |2.33333333333333|1.0434983894999 |2.0869967789998  |1.86052101883813  |0.832050294337844|1.66410058867569|
|1      |4      |1    |2    |3               |1.34164078649987|2.68328157299975 |2                 |0.894427190999916|1.78885438199983|
|1      |5      |1    |2    |3.66666666666667|1.63978318349985|3.27956636699969 |2.0761369963435   |0.928476690885259|1.85695338177052|
|1      |6      |1    |2    |4.33333333333333|1.93792558049982|3.87585116099964 |2.12132034355964  |0.948683298050514|1.89736659610103|
|1      |7      |1    |2    |5               |2.23606797749979|4.47213595499958 |2.15003290891883  |0.961523947640823|1.92304789528165|
|1      |1      |1    |3    |1               |0.316227766016838|0.948683298050514|1                 |0.316227766016838|0.948683298050514|
|1      |2      |1    |3    |1.75            |0.553398590529466|1.6601957715884  |1.75411603861406  |0.554700196225229|1.66410058867569|
|1      |3      |1    |3    |2.5             |0.790569415042095|2.37170824512628 |2.23606797749979  |0.707106781186548|2.12132034355964|
|1      |4      |1    |3    |3.25            |1.02774023955472|3.08322071866417 |2.5298221281347   |0.8              |2.4             |
|1      |5      |1    |3    |4               |1.26491106406735|3.79473319220206 |2.7116307227332   |0.857492925712544|2.57247877713763|
|1      |6      |1    |3    |4.75            |1.50208188857998|4.50624566573994 |2.82842712474619  |0.894427190999916|2.68328157299975|
|1      |7      |1    |3    |5.5             |1.73925271309261|5.21775813927783 |2.9065917948809   |0.919145030018058|2.75743509005417|
|1      |1      |2    |1    |1               |0.894427190999916|0.447213595499958|1                 |0.894427190999916|0.447213595499958|
|1      |2      |2    |1    |1.33333333333333|1.19256958799989|0.596284793999944|1.08465228909328  |0.970142500145332|0.485071250072666|
|1      |3      |2    |1    |1.66666666666667|1.49071198499986|0.74535599249993 |1.10282193314071  |0.986393923832144|0.493196961916072|
|1      |4      |2    |1    |2               |1.78885438199983|0.894427190999916|1.10940039245046  |0.992277876713668|0.496138938356834|
|1      |5      |2    |1    |2.33333333333333|2.0869967789998 |1.0434983894999  |1.11248539872496  |0.995037190209989|0.497518595104995|
|1      |6      |2    |1    |2.66666666666667|2.38513917599978|1.19256958799989 |1.11417202906231  |0.99654575824488 |0.49827287912244|
|1      |7      |2    |1    |3               |2.68328157299975|1.34164078649987 |1.11519272878504  |0.997458699830735|0.498729349915368|
|1      |1      |3    |1    |1               |0.948683298050514|0.316227766016838|1                 |0.948683298050514|0.316227766016838|
|1      |2      |3    |1    |1.25            |1.18585412256314|0.395284707521047|1.03975048982007  |0.986393923832144|0.328797974610715|
|1      |3      |3    |1    |1.5             |1.42302494707577|0.474341649025257|1.04764544365437  |0.993883734673619|0.33129457822454|
|1      |4      |3    |1    |1.75            |1.6601957715884 |0.553398590529466|1.05045146287778  |0.99654575824488 |0.33218191941496|
|1      |5      |3    |1    |2               |1.89736659610103|0.632455532033676|1.05175790477918  |0.997785157856609|0.33259505261887|
|1      |6      |3    |1    |2.25            |2.13453742061366|0.711512473537885|1.05246962316844  |0.998460353205412|0.332820117735138|
|1      |7      |3    |1    |2.5             |2.37170824512628|0.790569415042095|1.05289946579333  |0.998868137724438|0.332956045908146|
</details>

The "live" code sometimes results in components for the velocity that are greater than the maximum set in the same direction (i.e.: it may result in `newxmove` greater than `speed_x`). This different method should normalize everything in such a way that it never happens.

### CODE REVIEW

**Comment by ivan-mogilko** [comment]
I have to comment on this for clarification, and for the record.

This is about Character and Object movement speed. The problem noticed by @AlanDrake was that when there are distinct X and Y speeds, the calculations used to make a final speed is not very logical, and potentially results in final speed being higher than the maximal of X and Y, which normally should not happen.

@MatteoPiovanelli-Laser suggested that the correct calculation should, in geometric terms, limit the final speed by a "ellipse" where set X and Y speeds are ellipses axes, and the final speed's direction is matching the direction of the movement. In other words, we imagine a speed vector which length is always restricted by certain ellipse.

image

**Comment by AlanDrake** [comment]
With these changes, walking speed is closer to what I'd expect, but why does it face up while walking down when vertical speed is 1?

**Comment by ivan-mogilko** [comment]
I think this looks fine and may be merged.

File: Engine/ac/route_finder_impl.cpp [hunk 1]
```diff
@@ -218,6 +220,9 @@ void calculate_move_stage(MoveList * mlsp, int index)
   //fixed newxmove = useMoveSpeed * fcos(angl);
   float newxmove = useMoveSpeed * cos(angl);
 
+  // validate that the computed movement isn't larger than the set maxima
+  assert(newxmove <= move_speed_x && newymove <= move_speed_y);
```
Comment (ivan-mogilko) [inline Engine/ac/route_finder_impl.cpp:224]: I think there's a mistake in this assert: newxmove and newymove are signed values, so you might need to take std::abs(newxmove).

Comment (ivan-mogilko) [reply Engine/ac/route_finder_impl.cpp:224]: Oh sorry, I misread where they are getting negated.

**Pull request merged**